Add sorting by team_count, as well as secondary sort (by name).

cahrens opened the pull request:
TNL-3012

Adds sorting of topics by team_count, and also ensures a secondary sort (in the case of ties) by name.

@peter-fogg and @bderusha can you please review?

cahrens commented on lms/djangoapps/teams/tests/test_views.py:
```diff
@@ -363,7 +363,7 @@ def test_filter_text_search(self):
         (None, 200, ['Nuclear Team', u'sólar team', 'Wind Team']),
         ('name', 200, ['Nuclear Team', u'sólar team', 'Wind Team']),
-        ('open_slots', 200, ['Wind Team', u'sólar team', 'Nuclear Team']),
+        ('open_slots', 200, ['Wind Team', 'Nuclear Team', u'sólar team']),
```
Note to reviewers: Added secondary sort by name, per conversation with @explorerleslie. "Nuclear Team" and "solar team" have the same number of team members.

bderusha replied:
It may be worth adding a comment to that effect in the code for future maintainers of this file

cahrens commented on lms/djangoapps/teams/views.py:
```diff
@@ -315,14 +314,15 @@ def get(self, request):
         elif order_by_input == 'open_slots':
             queryset = queryset.annotate(team_size=Count('users'))
-            order_by_field = 'team_size'
+            queryset = queryset.order_by('team_size', 'lower_name')
```
Note to reviewers: this sorts by team_size as primary, lower_name as secondary.

cahrens commented on lms/djangoapps/teams/serializers.py:
```diff
@@ -160,3 +162,6 @@ def __init__(self, *args, **kwargs):
         for topic in self.data['results']:
             topic['team_count'] = topics_to_team_count.get(topic['id'], 0)
+
+        if 'sort_order' in self.context and self.context['sort_order'] == 'team_count':
```
Note to reviewers: sort_order was always being included in the context, but was ignored by this method. I chose to keep alphabetical sorting out of this method (as it was before), but am doing the team_count primary sort here.

Python documentation recommends doing two separate sorts in this sort of case instead of providing a more complicated comparator function that can handle the primary/secondary sorting as a single operation. 
https://docs.python.org/2/howto/sorting.html#sort-stability-and-complex-sorts

dan-f replied:
I think this approach may be reversed. It looks to me like the alphabetical sort is currently the primary, while the secondary sort is `team_count`: first we sort all the topics on the course module alphabetically in `get_alphabetical_topics`, we then cut off a page of results, and hand it to the numerical `team_count` sort which only sorts within the page.

 This will sort the current page by `team_count`, but if the current page itself only includes topics with low `team_count`s, then we're not really sorting by team count.

peter-fogg replied:
The order is correct, in that the sort by `team_count` is taking precedence over the alphabetical sort. I think the best solution is the sort by both `name` and `team_count` before paginating. Since this is a little confusing, could you add a test that sorts a list of topics where two topics `aardvarks` and `zebras` have the same team count, but would get put on different pages when sorting alphabetically?

cahrens commented on lms/djangoapps/teams/tests/test_views.py:
```diff
@@ -597,10 +597,18 @@ def test_without_course_id(self):
     @ddt.data(
         (None, 200, ['Coal Power', 'Nuclear Power', u'sólar power', 'Wind Power'], 'name'),
         ('name', 200, ['Coal Power', 'Nuclear Power', u'sólar power', 'Wind Power'], 'name'),
+        ('team_count', 200, ['Nuclear Power', u'sólar power', 'Coal Power', 'Wind Power'], 'team_count'),
```
Note to reviewers: "Coal Power" and "Wind Power" have the same team_count, so this is also verifying the secondary sort by name.

bderusha replied:
It may be worth adding a comment to that effect in the code for future maintainers of this file

peter-fogg commented on lms/djangoapps/teams/serializers.py:
```diff
@@ -160,3 +162,6 @@ def __init__(self, *args, **kwargs):
             topic['team_count'] = topics_to_team_count.get(topic['id'], 0)
+
+        if 'sort_order' in self.context and self.context['sort_order'] == 'team_count':
+            list.sort(self.data['results'], key=lambda t: t['team_count'], reverse=True)
```
Minor nit: could you use `self.data['results'].sort( ... )` here? I usually find that notation a bit more readable.

cahrens commented on lms/djangoapps/teams/tests/test_serializers.py:
```diff
@@ -181,3 +187,28 @@ def test_scoped_within_course(self):
+
+    def test_topics_with_team_counts_prepopulated(self):
+        """
+        Verify that if team_count is pre-populated, there are no additional SQL queries..
```
Nit: extra "." (I'll clean this up when addressing review feedback).

cahrens commented on lms/djangoapps/teams/views.py:
```diff
@@ -403,6 +407,28 @@ def post(self, request):
             return Response(CourseTeamSerializer(team).data)
 
 
+class IsEnrolledOrIsStaff(permissions.BasePermission):
```
Note to reviewers: just a small amount of cleanup following @efischer19's PR, #9289.

efischer19 replied:
Thanks for fixing this! To clarify, this is in response to @nedbat's comments on https://github.com/edx/edx-platform/pull/9323

cahrens commented:
@peter-fogg and @bderusha I think that this is now ready for review.

bderusha commented on lms/djangoapps/teams/views.py:
```diff
@@ -620,22 +623,37 @@ def get(self, request):
+        # Always sort alphabetically, as it will be used as secondary sort
+        # in the case of "team_count".
+        topics = get_alphabetical_topics(course_module)
+        if ordering == 'team_count':
```
There doesn't seem to be much precedent for this, but I wonder if it makes sense (as this method continues to grow) to include some private methods on `TopicListView` that can help make the `get` method more readable (It's over 60 lines long with this addition).  Not a deal breaker, but it may be worth having the discussion at least.

cahrens replied:
I actually find it quite readable as-is, so I'm not included to split it up. Thanks for the suggestion though-- something to keep in mind in general.

peter-fogg commented:
:+1:

cahrens commented:
@bderusha I added the 2 comments you requested, and went ahead and squashed at the same time since it is pretty easy to find them in the test file.

Can you take a look and possibly give me a thumbs-up? Thanks.

bderusha commented:
looks good to me :+1:

cahrens commented on lms/djangoapps/teams/views.py:
```diff
@@ -322,14 +327,15 @@ def get(self, request):
             )
 
         queryset = CourseTeam.objects.filter(**result_filter)
+        # We will always use name as either a primary or secondary sort.
```
Note that although @explorerleslie and I talked about this before she left, secondary sort by name is not consistent with what is in https://openedx.atlassian.net/browse/TNL-1937. I have asked Lou for clarification and will change this if necessary in my next PR, which will add sorting by last_activity_at to this endpoint.
